### PR TITLE
Implement Pectra Devnet 4 changes for requests and 7702

### PR DESCRIPTION
**Pull request by gurukamath**
This PR incorporates the following updates to the Prague EIPs

EIP-7002: add logging to system contract Update EIP-7002: add logging to system contract EIPs#8890
EIP-7251: max eb, change request to flat encoding Update EIP-7251: change request to flat encoding EIPs#8857
EIP-6110: deposits, change request to flat encoding Update EIP-6110: change request to flat encoding EIPs#8856
EIP-7002: withdrawals, change request to flat encoding Update EIP-7002: change request to flat encoding EIPs#8855
EIP-7685: change requests hash to flat hash Update EIP-7685: change requests hash to flat hash EIPs#8854

Additionally, it implements the updates to EIP-7702

### What was wrong?
The version of the EIP-7685 and related EIPs is not in line with the latest discussions

### How was it fixed?
Update the EIPs

### CODE REVIEW

**Comment by gurukamath** [comment]
Tracking [PR](https://github.com/ethereum/execution-spec-tests/pull/832) for further updates

File: src/ethereum/prague/requests.py [hunk 1]
```diff
@@ -8,20 +8,11 @@
 [EIP-7685]: https://eips.ethereum.org/EIPS/eip-7685
 """
 
-from dataclasses import dataclass
-from typing import Tuple, Union
+from hashlib import sha256
```
Comment (SamWilsn) [inline src/ethereum/prague/requests.py:11]: Should we create a wrapper in `ethereum.crypto.hash` like we do for `keccak256`?

File: src/ethereum/prague/requests.py [hunk 2]
```diff
@@ -30,178 +21,53 @@
...
-    )
-
-    consolidation_requests: Tuple[Bytes, ...] = ()
-    for i in range(count_consolidation_requests):
-        start = i * CONSOLIDATION_REQUEST_LENGTH
-        consolidation_request = parse_consolidation_data(
-            evm_call_output[start : start + CONSOLIDATION_REQUEST_LENGTH]
-        )
-        consolidation_requests += (encode_request(consolidation_request),)
+    m = sha256()
```
Comment (SamWilsn) [inline src/ethereum/prague/requests.py:69]: Weird. We should make a note somewhere (probably in `ethereum.crypto.hash`) of exactly what algorithm this is. I believe it's SHA2-256, right?

File: src/ethereum/prague/utils/message.py [hunk 3]
```diff
@@ -78,6 +79,11 @@ def prepare_message(
     message: `ethereum.prague.vm.Message`
         Items containing contract creation or message call specific data.
     """
+    accessed_addresses = set()
+    accessed_addresses.add(caller)
+    accessed_addresses.update(PRE_COMPILED_CONTRACTS.keys())
+    accessed_addresses.update(preaccessed_addresses)
+
```
Comment (SamWilsn) [inline src/ethereum/prague/utils/message.py:86]: This change will need to be backported, yes?

File: src/ethereum/prague/vm/eoa_delegation.py [hunk 4]
```diff
@@ -48,6 +50,25 @@ def is_valid_delegation(code: bytes) -> bool:
...
+    code: `bytes`
+        The code to get the address from.
+
+    Returns
+    -------
+    address : `Optional[Address]`
+        The address of the delegated code.
+    """
+    if is_valid_delegation(code):
+        return Address(code[EOA_DELEGATION_MARKER_LENGTH:])
```
Comment (SamWilsn) [inline src/ethereum/prague/vm/eoa_delegation.py:68]: Is it possible for code to be longer than `EOA_DELEGATION_MARKER_LENGTH + 20`?

File: src/ethereum/prague/vm/eoa_delegation.py [hunk 5]
```diff
@@ -144,6 +165,9 @@ def set_delegation(message: Message, env: Environment) -> U256:
         if auth.chain_id not in (env.chain_id, U64(0)):
             continue
 
+        if auth.nonce >= 2**64 - 1:
```
Comment (SamWilsn) [inline src/ethereum/prague/vm/eoa_delegation.py:168]: Should we use `U64.MAX_VALUE`?

Comment (SamWilsn) [inline src/ethereum/prague/vm/eoa_delegation.py:168]: Do we do the `assert auth.nonce < 2**64` or else the tx is invalid? Probably, and I just haven't come across it yet.

File: src/ethereum/prague/requests.py (hunk 1 above)
Comment (gurukamath) [reply src/ethereum/prague/requests.py:11]: Unless we plan on moving this whole function (for computing the hash of a list of bytes) to `ethereum.crypto`, we will need it here

File: src/ethereum/prague/fork.py [hunk 6]
```diff
@@ -820,7 +814,83 @@ def apply_body(
...
+        root(receipts_trie),
+        block_logs_bloom,
+        state_root(state),
+        root(withdrawals_trie),
+        blob_gas_used,
+        requests_hash,
+    )
+
+
+def process_requests(
```
[the remark this answers is not in the document]
Comment (gurukamath) [reply src/ethereum/prague/fork.py:845]: Hmm. I think `requests` has come to be accepted as refering to the general purpose execution layer requests from EIP-7685. However, I see the point about it being confusing to non-EL audience. Switched it to `process_general_purpose_requests`

File: src/ethereum/prague/requests.py [hunk 7]
```diff
@@ -30,178 +21,53 @@
...
-    elif data.startswith(WITHDRAWAL_REQUEST_TYPE):
-        return rlp.decode_to(WithdrawalRequest, data[1:])
-    elif data.startswith(CONSOLIDATION_REQUEST_TYPE):
-        return rlp.decode_to(ConsolidationRequest, data[1:])
-    else:
-        raise Exception("Unknown request type")
-
-
-def parse_deposit_data(data: Bytes) -> DepositRequest:
+def parse_deposit_data(data: Bytes) -> Bytes:
```
[the remark this answers is not in the document]
Comment (gurukamath) [reply src/ethereum/prague/requests.py:26]: Updated

File: src/ethereum/prague/requests.py (hunk 2 above)
Comment (gurukamath) [reply src/ethereum/prague/requests.py:69]: Added a note to the function doc string

File: src/ethereum/prague/utils/message.py (hunk 3 above)
Comment (gurukamath) [reply src/ethereum/prague/utils/message.py:86]: Yes. Added to #1026

File: src/ethereum/prague/vm/eoa_delegation.py (hunk 4 above)
Comment (gurukamath) [reply src/ethereum/prague/vm/eoa_delegation.py:68]: Not for valid delegations. One of the checks we make is that the code size is `EOA_DELEGATION_MARKER_LENGTH + 20`

File: src/ethereum/prague/vm/eoa_delegation.py (hunk 5 above)
Comment (gurukamath) [reply src/ethereum/prague/vm/eoa_delegation.py:168]: Updated

Comment (gurukamath) [reply src/ethereum/prague/vm/eoa_delegation.py:168]: We don't do this explicitly but is implied in the fact that auth.nonce is U64.

**Review by SamWilsn — APPROVED** [review]

**Review by petertdavies — APPROVED** [review]

**Pull request merged**